// ngx-infinite-scroller/esm5/src/app/scrolling-strategy/scrolling-to-bottom.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { filter } from 'rxjs/operators';
import { StrategyBase } from './strategy-base';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
var ScrollingToBottom = /** @class */ (function (_super) {
    tslib_1.__extends(ScrollingToBottom, _super);
    function ScrollingToBottom(directive, state) {
        return _super.call(this, directive, state) || this;
    }
    /**
     * @param {?} scrollPairChanged
     * @return {?}
     */
    ScrollingToBottom.prototype.scrollDirectionChanged = /**
     * @param {?} scrollPairChanged
     * @return {?}
     */
    function (scrollPairChanged) {
        var _this = this;
        return scrollPairChanged.pipe(filter(function (scrollPositions) {
            return _super.prototype.wasScrolledDown.call(_this, scrollPositions[0], scrollPositions[1]);
        }));
    };
    /**
     * @param {?} scrollDirectionChanged
     * @return {?}
     */
    ScrollingToBottom.prototype.scrollRequestZoneChanged = /**
     * @param {?} scrollDirectionChanged
     * @return {?}
     */
    function (scrollDirectionChanged) {
        var _this = this;
        return scrollDirectionChanged.pipe(filter(function (scrollPositions) {
            return _super.prototype.isScrollDownEnough.call(_this, scrollPositions[1], _this.directive.scrollDownPercentilePositionTrigger);
        }));
    };
    /**
     * @return {?}
     */
    ScrollingToBottom.prototype.askForUpdate = /**
     * @return {?}
     */
    function () {
        this.directive.onScrollDown.next();
    };
    /**
     * @return {?}
     */
    ScrollingToBottom.prototype.setInitialScrollPosition = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.TOP);
        this.directive.scrollTo(initialScrollPositionValue);
    };
    /**
     * @return {?}
     */
    ScrollingToBottom.prototype.setPreviousScrollPosition = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var prevScrollPosition = this.state.previousScrollTop;
        this.directive.scrollTo(prevScrollPosition);
    };
    return ScrollingToBottom;
}(StrategyBase));
export { ScrollingToBottom };
