// ngx-infinite-scroller/esm5/src/app/scrolling-strategy/scrolling-to-both.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { tap, filter } from 'rxjs/operators';
import { StrategyBase } from './strategy-base';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
var ScrollingToBoth = /** @class */ (function (_super) {
    tslib_1.__extends(ScrollingToBoth, _super);
    function ScrollingToBoth(directive, state) {
        return _super.call(this, directive, state) || this;
    }
    /**
     * @param {?} scrollPairChanged
     * @return {?}
     */
    ScrollingToBoth.prototype.scrollDirectionChanged = /**
     * @param {?} scrollPairChanged
     * @return {?}
     */
    function (scrollPairChanged) {
        return scrollPairChanged;
    };
    /**
     * @param {?} scrollDirectionChanged
     * @return {?}
     */
    ScrollingToBoth.prototype.scrollRequestZoneChanged = /**
     * @param {?} scrollDirectionChanged
     * @return {?}
     */
    function (scrollDirectionChanged) {
        var _this = this;
        return scrollDirectionChanged.pipe(filter(function (scrollPositions) {
            return (_super.prototype.isScrollUpEnough.call(_this, scrollPositions[1], _this.directive.scrollUpPercentilePositionTrigger) || _super.prototype.isScrollDownEnough.call(_this, scrollPositions[1], _this.directive.scrollDownPercentilePositionTrigger));
        }), tap(function (scrollPositions) {
            _this.scrolledUp = _super.prototype.wasScrolledUp.call(_this, scrollPositions[0], scrollPositions[1]);
        }));
    };
    /**
     * @return {?}
     */
    ScrollingToBoth.prototype.askForUpdate = /**
     * @return {?}
     */
    function () {
        if (this.scrolledUp) {
            this.directive.onScrollUp.next();
        }
        else {
            this.directive.onScrollDown.next();
        }
    };
    /**
     * @return {?}
     */
    ScrollingToBoth.prototype.setInitialScrollPosition = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.MIDDLE);
        this.directive.scrollTo(initialScrollPositionValue);
    };
    /**
     * @return {?}
     */
    ScrollingToBoth.prototype.setPreviousScrollPosition = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var prevScrollPosition;
        if (this.scrolledUp) {
            prevScrollPosition = this.state.previousScrollTop +
                (this.state.scrollHeight - this.state.previousScrollHeight);
        }
        else {
            prevScrollPosition = this.state.previousScrollTop;
        }
        this.directive.scrollTo(prevScrollPosition);
    };
    return ScrollingToBoth;
}(StrategyBase));
export { ScrollingToBoth };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollingToBoth.prototype.scrolledUp;
}
