// ngx-infinite-scroller/esm5/src/app/ngx-infinite-scroller.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgxInfiniteScrollerDirective } from './ngx-infinite-scroller.directive';
import { DirectiveStateService } from './directive-state.service';
var NgxInfiniteScrollerModule = /** @class */ (function () {
    function NgxInfiniteScrollerModule() {
    }
    NgxInfiniteScrollerModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        NgxInfiniteScrollerDirective
                    ],
                    imports: [],
                    exports: [
                        NgxInfiniteScrollerDirective
                    ],
                    providers: [
                        DirectiveStateService
                    ],
                    bootstrap: []
                },] }
    ];
    return NgxInfiniteScrollerModule;
}());
export { NgxInfiniteScrollerModule };
