// ngx-infinite-scroller/esm2015/src/app/directive-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class DirectiveStateService {
    /**
     * @return {?}
     */
    get scrollTop() {
        return this._el.nativeElement.scrollTop;
    }
    /**
     * @return {?}
     */
    get scrollHeight() {
        return this._el.nativeElement.scrollHeight;
    }
    /**
     * @return {?}
     */
    get clientHeight() {
        return this._el.nativeElement.clientHeight;
    }
    /**
     * @return {?}
     */
    get initMode() {
        return this._initMode;
    }
    /**
     * @param {?} initMode
     * @return {?}
     */
    set initMode(initMode) {
        this._initMode = initMode;
    }
    /**
     * @return {?}
     */
    get scrollStreamActive() {
        return this._scrollStreamActive;
    }
    /**
     * @param {?} active
     * @return {?}
     */
    set scrollStreamActive(active) {
        this._scrollStreamActive = active;
    }
    /**
     * @return {?}
     */
    get previousScrollPositionpUpdated() {
        return this._previousScrollPositionpUpdated;
    }
    /**
     * @param {?} previousScrollPositionpUpdated
     * @return {?}
     */
    set previousScrollPositionpUpdated(previousScrollPositionpUpdated) {
        this._previousScrollPositionpUpdated = previousScrollPositionpUpdated;
    }
    /**
     * @return {?}
     */
    get previousScrollTop() {
        return this._previousScrollTop;
    }
    /**
     * @return {?}
     */
    get previousScrollHeight() {
        return this._previousScrollHeight;
    }
    /**
     * @param {?} params
     * @return {?}
     */
    setup(params) {
        this._el = params.el;
        this._initMode = params.initMode;
        this._scrollStreamActive = params.scrollStreamActive;
        this._previousScrollPositionpUpdated = params.previousScrollPositionpUpdated;
        this.updatePreviousScrollTop();
        this.updatePreviousScrollHeight();
    }
    /**
     * @return {?}
     */
    updatePreviousScrollTop() {
        this._previousScrollTop = this._el.nativeElement.scrollTop;
    }
    /**
     * @return {?}
     */
    updatePreviousScrollHeight() {
        this._previousScrollHeight = this._el.nativeElement.scrollHeight;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    DirectiveStateService.prototype._el;
    /**
     * @type {?}
     * @private
     */
    DirectiveStateService.prototype._initMode;
    /**
     * @type {?}
     * @private
     */
    DirectiveStateService.prototype._scrollStreamActive;
    /**
     * @type {?}
     * @private
     */
    DirectiveStateService.prototype._previousScrollPositionpUpdated;
    /**
     * @type {?}
     * @private
     */
    DirectiveStateService.prototype._previousScrollTop;
    /**
     * @type {?}
     * @private
     */
    DirectiveStateService.prototype._previousScrollHeight;
}
