// @thisissoon/angular-infinite-scroll/esm2015/app/infinite-scroll/infinite-scroll.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, EventEmitter, Input, Output, } from '@angular/core';
import { Subject } from 'rxjs';
import { debounceTime, takeUntil, filter, map } from 'rxjs/operators';
import * as events from './shared/events';
/**
 * \@example
 * ```html
 * <div
 *   class="foo"
 *   snInfiniteScroll
 *   (scrollEnd)="onScrollEnd()"
 *   [offset]="100"
 *   [debounce]="123"
 *   [disabled]="disabled">
 * </div>
 * ```
 *
 */
export class InfiniteScrollDirective {
    constructor() {
        /**
         * Event that will be triggered when user has scrolled to
         * bottom of the element
         */
        this.scrollEnd = new EventEmitter();
        /**
         * An offset from the bottom of the element to trigger
         * `scrollEnd` event
         */
        this.offset = 0;
        /**
         * Specify debounce duration in ms
         */
        this.debounce = 100;
        /**
         * If true then `scrollEnd` event should NOT be emitted
         */
        this.disabled = false;
        /**
         * Emits a new value on element scroll event
         */
        this.scroll$ = new Subject();
        /**
         * Completes on component destroy lifecycle event
         * used to unsubscribe from infinite observables
         *
         */
        this.ngUnsubscribe$ = new Subject();
    }
    /**
     * Subscribe to `scroll$` observable and emit `scrollEnd` event
     * when element scroll position is at the end of the element
     * @return {?}
     */
    ngAfterViewInit() {
        this.scroll$
            .pipe(takeUntil(this.ngUnsubscribe$), debounceTime(this.debounce), map(scroll => {
            /** @type {?} */
            const y = scroll.y + this.offset;
            return { y, height: scroll.height };
        }), filter(() => !this.disabled), filter(scroll => scroll.y >= scroll.height))
            .subscribe(() => this.scrollEnd.emit());
    }
    /**
     * On element scroll event emit next `scroll$` observable value
     * @param {?} scrollY
     * @param {?} scrollheight
     * @param {?} offsetHeight
     * @return {?}
     */
    onScroll(scrollY, scrollheight, offsetHeight) {
        /** @type {?} */
        const height = scrollheight;
        /** @type {?} */
        const y = scrollY + offsetHeight;
        this.scroll$.next({ y, height });
    }
    /**
     * trigger `ngUnsubscribe` complete on
     * component destroy lifecycle hook
     * @return {?}
     */
    ngOnDestroy() {
        this.ngUnsubscribe$.next();
        this.ngUnsubscribe$.complete();
    }
}
InfiniteScrollDirective.decorators = [
    { type: Directive, args: [{
                selector: '[snInfiniteScroll]',
            },] }
];
InfiniteScrollDirective.propDecorators = {
    scrollEnd: [{ type: Output }],
    offset: [{ type: Input }],
    debounce: [{ type: Input }],
    disabled: [{ type: Input }],
    onScroll: [{ type: HostListener, args: [events.eventScroll, events.eventPathScroll,] }]
};
if (false) {
    /**
     * Event that will be triggered when user has scrolled to
     * bottom of the element
     * @type {?}
     */
    InfiniteScrollDirective.prototype.scrollEnd;
    /**
     * An offset from the bottom of the element to trigger
     * `scrollEnd` event
     * @type {?}
     */
    InfiniteScrollDirective.prototype.offset;
    /**
     * Specify debounce duration in ms
     * @type {?}
     */
    InfiniteScrollDirective.prototype.debounce;
    /**
     * If true then `scrollEnd` event should NOT be emitted
     * @type {?}
     */
    InfiniteScrollDirective.prototype.disabled;
    /**
     * Emits a new value on element scroll event
     * @type {?}
     */
    InfiniteScrollDirective.prototype.scroll$;
    /**
     * Completes on component destroy lifecycle event
     * used to unsubscribe from infinite observables
     *
     * @type {?}
     * @private
     */
    InfiniteScrollDirective.prototype.ngUnsubscribe$;
}
