// ngx-infinite-scroller/esm2015/src/app/scroll-height-listener/scroll-height-listener.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ScrollHeightListener {
    /**
     * @param {?} directive
     * @param {?} state
     */
    constructor(directive, state) {
        this.directive = directive;
        this.state = state;
        this.DEFAULT_REQUEST_TIMEOUT = 30000;
    }
    /**
     * @return {?}
     */
    start() {
        this.listener = window.requestAnimationFrame(this.listen.bind(this));
        if (!this.httpRequestTimeout) {
            this.httpRequestTimeout = setTimeout(() => {
                this.stopIfRequestTimeout();
            }, this.DEFAULT_REQUEST_TIMEOUT);
        }
    }
    /**
     * @return {?}
     */
    stop() {
        window.cancelAnimationFrame(this.listener);
        clearTimeout(this.httpRequestTimeout);
        this.httpRequestTimeout = null;
    }
    /**
     * @private
     * @return {?}
     */
    listen() {
        if (this.state.previousScrollHeight !== this.state.scrollHeight) {
            this.stop();
            this.directive.onScrollbarHeightChanged();
        }
        else {
            this.start();
        }
    }
    /**
     * @private
     * @return {?}
     */
    stopIfRequestTimeout() {
        if (!this.state.previousScrollPositionpUpdated) {
            this.stop();
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollHeightListener.prototype.DEFAULT_REQUEST_TIMEOUT;
    /**
     * @type {?}
     * @private
     */
    ScrollHeightListener.prototype.listener;
    /**
     * @type {?}
     * @private
     */
    ScrollHeightListener.prototype.httpRequestTimeout;
    /**
     * @type {?}
     * @private
     */
    ScrollHeightListener.prototype.directive;
    /**
     * @type {?}
     * @private
     */
    ScrollHeightListener.prototype.state;
}
