// ngx-infinite-scroller/esm2015/src/app/scrolling-strategy/strategy-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import isNumber from 'is-number';
import { InitialScrollPosition } from '../enum/initial-scroll-position-type.enum';
/**
 * @abstract
 */
export class StrategyBase {
    /**
     * @param {?} directive
     * @param {?} state
     */
    constructor(directive, state) {
        this.directive = directive;
        this.state = state;
    }
    /**
     * @protected
     * @param {?} prevPos
     * @param {?} currentPos
     * @return {?}
     */
    wasScrolledDown(prevPos, currentPos) {
        return prevPos.scrollTop < currentPos.scrollTop;
    }
    /**
     * @protected
     * @param {?} prevPos
     * @param {?} currentPos
     * @return {?}
     */
    wasScrolledUp(prevPos, currentPos) {
        return !this.wasScrolledDown(prevPos, currentPos);
    }
    /**
     * @protected
     * @param {?} pos
     * @param {?} scrollPositionTrigger
     * @return {?}
     */
    isScrollDownEnough(pos, scrollPositionTrigger) {
        return ((pos.scrollTop + pos.clientHeight) / pos.scrollHeight) > (scrollPositionTrigger / 100);
    }
    /**
     * @protected
     * @param {?} pos
     * @param {?} scrollPositionTrigger
     * @return {?}
     */
    isScrollUpEnough(pos, scrollPositionTrigger) {
        return (pos.scrollTop / pos.scrollHeight) < (scrollPositionTrigger / 100);
    }
    /**
     * @protected
     * @param {?} defaultScrollPosition
     * @return {?}
     */
    getInitialScrollPositionValue(defaultScrollPosition) {
        const { initialScrollPosition } = this.directive;
        if (isNumber(initialScrollPosition)) {
            return Number(initialScrollPosition);
        }
        /** @type {?} */
        const initialScrollPositions = this.getInitialScrollPositions();
        if (initialScrollPosition === InitialScrollPosition.DEFAULT) {
            return initialScrollPositions[defaultScrollPosition];
        }
        return initialScrollPositions[initialScrollPosition];
    }
    /**
     * @private
     * @return {?}
     */
    getInitialScrollPositions() {
        const { scrollHeight, clientHeight } = this.state;
        return {
            [InitialScrollPosition.TOP]: 0,
            [InitialScrollPosition.MIDDLE]: scrollHeight / 2 - clientHeight / 2,
            [InitialScrollPosition.BOTTOM]: scrollHeight,
        };
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    StrategyBase.prototype.directive;
    /**
     * @type {?}
     * @protected
     */
    StrategyBase.prototype.state;
}
